src/mid.cpp

```diff
@@ -941,21 +955,50 @@ void CmidPlayer::rewind(int subsong)
                 adlib_style=LUCAS_STYLE|MIDI_STYLE;
                 //note: no break, we go right into midi headers...
             case FILE_MIDI:
+            case FILE_MIDItype1:
                 if (type != FILE_LUCAS)
                     tins=128;
-                getnext(11);  /*skip header*/
+                /* "Mthd"              header-chunk
+                 * 0x00 0x00 0x00 0x06 header size
+                 * 0x00 0x0n           midi file type: 0=single-track format, 1=multiple-track format, 2=multiple-song format
+                 * 0xnn 0xnn           track-number
+                 * 0xnn 0xnn           tempo
+                 */
+                getnext(7);
+                i=getnext(2);
+                type=(i==1)?FILE_MIDItype1:FILE_MIDI;
+		midiprintf ("type:%ld\n", i);
+                getnext(2); /* skip track-number */
                 deltas=getnext(2);
                 midiprintf ("deltas:%ld\n",deltas);
-                getnext(4);
+
+	        for (i=0; i<16; i++)
+	        {
+                  ch[i].nshift=-13;
+                  ch[i].on=1;
+                }
 
                 curtrack=0;
-                track[curtrack].on=1;
-                track[curtrack].tend=getnext(4);
-                midiprintf ("tracklen:%lu\n",track[curtrack].tend);
-                //track[curtrack].tend += pos; // FIXME: length -> end position
-                if (track[curtrack].tend > flen) // no music after end of file
-                    track[curtrack].tend = flen;
-                track[curtrack].spos=pos;
+                while (((type==FILE_MIDI) && curtrack == 0) ||
+                       ((type==FILE_MIDItype1) && (curtrack < 16)))
+                {
+                    char s[4];
+                    s[0] = 0;
+                    readString(s, 4);
+                    midiprintf("Offset=0x%08lx\n", pos);
+                    midiprintf("trkHeader: %c%c%c%c\n", s[0], s[1], s[2], s[3]);
+                    if (s[0] != 'M' || s[1]!='T' || s[2]!='r' || s[3]!='k')
+                        break;
```

Comment:
It might be neater here to make `s` longer (`char s[5]`) then after `readString()` you can put `s[4] = 0;` to ensure it is a null-terminated string.  Then `s` will be a normal string so you can say `if (strcmp(s, "MTrk") == 0)` and likewise `trkHeader: %s`.

Reply:
Good idea. I was just following the coding-standard already used in this file.